Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+  `[-∞, +∞]`. This first version uses a `liminf`.-/
+noncomputable def coverEntropyInfUni (T : X → X) (F : Set X) (U : Set (X × X)) :=
+  atTop.liminf fun n : ℕ ↦ log (coverMincard T F U n) / n
+
+/-- The entropy of an entourage `U`, defined as the exponential rate of growth of the size of the
+  smallest `(n, U)`-refined cover of `F`. Takes values in the space fo extended real numbers
+  `[-∞, +∞]`. This second version uses a `limsup`.-/
+noncomputable def coverEntropySupUni (T : X → X) (F : Set X) (U : Set (X × X)) :=
+  atTop.limsup fun n : ℕ ↦ log (coverMincard T F U n) / n
+
+lemma coverEntropyInfUni_antitone (T : X → X) (F : Set X) :
+    Antitone (fun U : Set (X × X) ↦ coverEntropyInfUni T F U) :=
+  fun _ _ U_V ↦ (liminf_le_liminf) <| Eventually.of_forall
+    fun n ↦ monotone_div_right_of_nonneg (Nat.cast_nonneg' n)
+    <| log_monotone (ENat.toENNReal_mono (coverMincard_antitone T F n U_V))
+
+lemma coverEntropySupUni_antitone (T : X → X) (F : Set X) :
+    Antitone (fun U : Set (X × X) ↦ coverEntropySupUni T F U) :=
+  fun _ _ U_V ↦ (limsup_le_limsup) <| Eventually.of_forall
+    fun n ↦ monotone_div_right_of_nonneg (Nat.cast_nonneg' n)
+    <| log_monotone (ENat.toENNReal_mono (coverMincard_antitone T F n U_V))
+
+lemma coverEntropyInfUni_le_coverEntropySupUni (T : X → X) (F : Set X) (U : Set (X × X)) :
+    coverEntropyInfUni T F U ≤ coverEntropySupUni T F U := liminf_le_limsup
+
+@[simp]
+lemma coverEntropySupUni_empty {T : X → X} {U : Set (X × X)} :
+    coverEntropySupUni T ∅ U = ⊥ := by
+  suffices h : ∀ᶠ n : ℕ in atTop, log (coverMincard T ∅ U n) / n = ⊥ by
+    rw [coverEntropySupUni]
+    exact limsup_congr h ▸ limsup_const ⊥
+  · simp only [coverMincard_empty, ENat.toENNReal_zero, log_zero, eventually_atTop]
+    exact ⟨1, fun n n_pos ↦ bot_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)⟩
+
+@[simp]
+lemma coverEntropyInfUni_empty {T : X → X} {U : Set (X × X)} :
+    coverEntropyInfUni T ∅ U = ⊥ :=
+  eq_bot_mono (coverEntropyInfUni_le_coverEntropySupUni T ∅ U) coverEntropySupUni_empty
+
+lemma coverEntropyInfUni_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ coverEntropyInfUni T F U :=
+  (le_iInf fun n ↦ div_nonneg (log_coverMincard_nonneg T h U n) (Nat.cast_nonneg' n)).trans
+    iInf_le_liminf
+
+lemma coverEntropySupUni_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ coverEntropySupUni T F U :=
+  (coverEntropyInfUni_nonneg T h U).trans (coverEntropyInfUni_le_coverEntropySupUni T F U)
+
+lemma coverEntropyInfUni_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    coverEntropyInfUni T F univ = 0 := by
+  simp [coverEntropyInfUni, coverMincard_univ T h]
+
+lemma coverEntropySupUni_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    coverEntropySupUni T F univ = 0 := by
+  simp [coverEntropySupUni, coverMincard_univ T h]
+
+lemma coverEntropyInfUni_le_log_coverMincard_div {T : X → X} {F : Set X} (F_inv : MapsTo T F F)
```

Comment:
Wouldn't it be better to deduce this lemma from `coverEntropySupUni_le_log_coverMincard_div` and the trivial inequality? Or remove it altogether, as it's a trivial consequence of the other one?

Reply:
Good idea. I've just removed the `Inf` versions.

I've also proved directly `coverEntropyEnt_finite_of_isCompact_invariant` and removed the `Inf` version.